File: SegTreeLite.py
class SegTreeLite:
    def __init__(self, n: int):
        self.n = n
        self.sum = [0] * (n * 4)

    def _add(self, o: int, l: int, r: int, idx: int, val: int):
        if l == r:
            self.sum[o] += val
            return
        m = (l + r) // 2
        if idx <= m: self._add(o * 2, l, m, idx, val)
        else: self._add(o * 2 + 1, m + 1, r, idx, val)
        self.sum[o] = self.sum[o * 2] + self.sum[o * 2 + 1]

    def _query_sum(self, o: int, l: int, r: int, L: int, R: int):
        if L <= l and r <= R: return self.sum[o]
        sum = 0
        m = (l + r) // 2
        if L <= m: sum += self._query_sum(o * 2, l, m, L, R)
        if R > m: sum += self._query_sum(o * 2 + 1, m + 1, r, L, R)
        return sum
    
    def query_sum(self, L: int, R: int):
        '''
        Query sum in range [l,r)
        '''
        return self._query_sum(1,1,self.n,L+1,R)
    
    def add(self, idx: int, val: int):
        '''
        Increase arr[idx] by val
        '''
        return self._add(1, 1, self.n, idx+1, val)

File: SegTreeLite.cpp
#include <bits/stdc++.h>

using namespace std;

class SegTreeLite{

    public:
        int* _sum;
        int _n;

        SegTreeLite(int n){
            _n = n;
            _sum = (int *)malloc(n * 4 * sizeof(int));
            memset(_sum, 0, sizeof(int) * n * 4);
        }

    void _add(int o, int l, int r, int idx, int val) {
        if (l == r) {
            _sum[o] += val;
            return;
        }
        int m = (l + r) / 2;
        if (idx <= m) _add(o * 2, l, m, idx, val);
        else _add(o * 2 + 1, m + 1, r, idx, val);
        _sum[o] = _sum[o * 2] + _sum[o * 2 + 1];
    }

    int _query_sum(int o, int l, int r, int L, int R) {
        if (L <= l && r <= R) return _sum[o];
        int sum = 0;
        int m = (l + r) / 2;
        if (L <= m) sum += _query_sum(o * 2, l, m, L, R);
        if (R > m) sum += _query_sum(o * 2 + 1, m + 1, r, L, R);
        return sum;
    }
    
    int query_sum(int L,int R){
        /*
        Query sum in range [l,r)
        */
        return _query_sum(1,1,_n,L+1,R);
    }
    
    void add(int idx,int val){
        /*
        Increase arr[idx] by val
        */
        _add(1, 1, _n, idx+1, val);
    }
};